### ISSUE TITLE
[NodeJS] Fetch is is not a function for App Catalog / MsalFetchClient

### ISSUE THREAD

**Issue by JakeStanger** [pnp/pnpjs#1719]
### Category
- [ ] Enhancement
- [x] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [2.4.0 commonjs]

```
λ node --version                                          
v14.15.4
```

Please specify what version(s) of SharePoint you are targeting: [SPO]

### Expected / Desired Behavior / Question

The request should succeed.

### Observed Behavior

The following error is produced:

```
(node:54284) UnhandledPromiseRejectionWarning: TypeError: safe_global_js_1.safeGlobal.fetch is not a function
    at FetchClient.fetch (/home/jake/Programming/Work/packages/common/net.ts:66:27)
    at SPHttpClient.retry (/home/jake/Programming/Work/packages/sp/sphttpclient.ts:121:24)
    at /home/jake/Programming/Work/packages/sp/sphttpclient.ts:143:19
    at new Promise (<anonymous>)
    at SPHttpClient.fetchRaw (/home/jake/Programming/Work/packages/sp/sphttpclient.ts:141:16)
    at /home/jake/Programming/Work/packages/sp/sphttpclient.ts:214:35
    at step (/home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:143:27)
    at Object.next (/home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:124:57)
    at /home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:117:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:113:16)
    at SPHttpClient._digestCache (/home/jake/Programming/Work/packages/sp/sphttpclient.ts:196:33)
    at SPHttpClient.<anonymous> (/home/jake/Programming/Work/packages/sp/sphttpclient.ts:78:39)
    at step (/home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:143:27)
    at Object.next (/home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:124:57)
    at /home/jake/Programming/Work/spfx-deploy/node_modules/tslib/tslib.js:117:75
```

Alternatively with the following hack:
```ts
import fetch from 'node-fetch';
(global as any).fetch = fetch;
```

You get:
```TypeError: Only absolute URLs are supported```

### Steps to Reproduce

Import sp:
```ts
import { sp } from '@pnp/sp-commonjs/presets/all';
```

```ts
// Gets an isolated client using `MsalFetchClient` with certificate authentication
// https://pnp.github.io/pnpjs/authentication/server-nodejs/#call-sharepoint
const spClient = await getSpClient(site);

// These succeed and log correctly
console.log(await spClient.web.get());
console.log(await spClient.web.lists.get());

const catalog = spClient.web.getAppCatalog();
const tenantCatalog = await spClient.getTenantAppCatalogWeb();

// Each of these fails with the above error
await catalog.add(artifact.name, artifact.file, true);
console.log(await catalog.get())
console.log(await tenantCatalog.get())
```

<details><summary>`getSpClient` for reference</summary>

```ts
async function getSpClient(site: Site) {
  const spClient = await sp.createIsolated();

  const tenant = await prisma.tenant.findUnique({
    where: { subdomain: site.tenantSubdomain },
  });

  const privateKey = await getPrivateKey(tenant.subdomain);

  spClient.setup({
    sp: {
      baseUrl: `https://${tenant.subdomain}.sharepoint.com/${site.url}`,
      fetchClientFactory: () => {
        return new MsalFetchClient(
          {
            auth: {
              authority: `https://login.microsoftonline.com/${tenant.id}/`,
              clientCertificate: {
                thumbprint: tenant.thumbprint,
                privateKey,
              },
              clientId: tenant.clientId,
            },
          },
          [`https://${tenant.subdomain}.sharepoint.com/.default`]
        ); // you must set the scope for SharePoint access
      },
    },
  });

  return spClient;
}
```

</details>

Let me know if there's anything else you need to know or want me to test.

**Comment by patrick-rodgers** [comment]
Where are you getting `MsalFetchClient`?d Based on the error it doesn't appear to be [the node one](https://pnp.github.io/pnpjs/authentication/server-nodejs/#msal)?

**Comment by JakeStanger** [comment]
I'm using the one from `node-commonjs`:

```ts
import { MsalFetchClient } from '@pnp/nodejs-commonjs';
```

**Comment by patrick-rodgers** [comment]
Looking at the code the stack trace you shared doesn't seem to match the code you've shared. The MsalFetchClient from the nodejs package doesn't as best I can tell use safe global at all. Are you able to share all the code? Everything in the nodejs package should be routed into the nodeFetch (fetch.js) and executed there.

**Comment by JakeStanger** [comment]
Interesting. I won't be able to share all the code but what I will do is firstly attack `yarn` and see if that's causing some bizarre issue, then try to put together a small repo which reproduces it. I'll try to get that to you tonight/tomorrow.

**Comment by patrick-rodgers** [comment]
Understood on sharing the code, no worries. But I would be curious to learn what's happening here to make sure we don't have some deeper issue I am not seeing at first glance. Thanks!

**Comment by JakeStanger** [comment]
Alright, I've extracted the bits of my code to put together a full working demo: https://github.com/JakeStanger/pnp-msal-bug-repro; you'll need to just fill out an `.env`file with auth details.

The biggest thing I've found is that I tried using the global `sp` client and it logged all the apps without a problem. The error was only thrown when using `sp.createIsolated()`. Hopefully that helps narrow it down a lot :)

**Comment by patrick-rodgers** [comment]
Thanks, much appreciated - we'll have a look.

**Comment by patrick-rodgers** [comment]
Ok, thanks to the repro I can see the bug and what appears to be happening is that for some reason the .setup call is correctly being applied to the DEFAULT runtime instead of the isolated. I am not sure why that's happening yet, but in essence the isolated runtime is not being updated as it should be. Working on figuring it out.

**Comment by patrick-rodgers** [comment]
Ok - two bugs here, one in your code and one in the library. In the repro you shared you applied the configuration to the global sp and not the isolated one - which was the source of the issue in my previous comment. That threw me off. After fixing that I found a bug in the getAppCatalog method in the lib that wasn't applying the runtime correctly. I have fixed that and will by part of 2.5.0 coming out soon. Thanks for helping us diagnose it!

**Comment by JakeStanger** [comment]
> In the repro you shared you applied the configuration to the global sp and not the isolated one

Oh you're right, that's correct in my actual project so I must have changed it when I was testing if the isolated runtime was at fault and forgotten to change it back. Sorry for throwing you off there.

Thanks for getting that fixed so quickly!

**Comment by patrick-rodgers** [comment]
All good - I didn't think of it because I had looked at the get client function shard in this thread and it was fine. Thanks!

**Comment by patrick-rodgers** [comment]
Closing this as fixed, just missed closing it with the last release. Fix went out in 2.5.0. Thanks!
